pkg/sources/s3/metrics.go

```diff
@@ -0,0 +1,104 @@
+package s3
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/trufflesecurity/trufflehog/v3/pkg/common"
+)
+
+// metricsCollector defines the interface for recording S3 scan metrics.
+type metricsCollector interface {
+	// Object metrics.
+
+	RecordObjectScanned(bucket string)
+	RecordObjectSkipped(bucket, reason string)
+	RecordObjectError(bucket string)
+
+	// Role metrics.
+
+	RecordRoleScanned(roleArn string)
+	RecordBucketForRole(roleArn string)
+}
+
+type collector struct {
+	objectsScanned *prometheus.CounterVec
+	objectsSkipped *prometheus.CounterVec
+	objectsErrors  *prometheus.CounterVec
+	rolesScanned   *prometheus.GaugeVec
+	bucketsPerRole *prometheus.GaugeVec
+}
+
+var metricsInstance metricsCollector
+
+func init() { metricsInstance = newS3MetricsCollector() }
+
+// getMetricsCollector returns the singleton metrics collector instance..
+func getMetricsCollector() metricsCollector { return metricsInstance }
```

Comment:
Why have a `get` function if `metricsInstance` is global?

Reply:
I was following the approach we use to access metrics in our cache package, but it doesn’t apply well here since the metrics are private. I'll adjust.